File: algorithms/2_medium/M/02542_maximum_subsequence_score/heap.cpp
class Solution {
public:
    static long long maxScore(const vector<int>& nums1, const vector<int>& nums2, const int k_int) {
        assert(nums1.size() == nums2.size() && k_int >= 1 && k_int <= std::ssize(nums1));

        const size_t len = nums1.size();
        const size_t k = static_cast<size_t>(k_int);
        auto combo = vector<pair<int, int>>(len);
        for (size_t i = 0; i < len; i += 1) {
            combo[i] = { nums2[i], nums1[i] };
        }
        ranges::sort(combo, [](const auto& a, const auto& b) {
            return a.first > b.first;
        });
        auto heap = priority_queue<int, vector<int>, std::greater<int>>();
        int64_t sum = 0;
        int64_t res = numeric_limits<int64_t>::min();

        for (size_t i = 0; i < len; i += 1) {
            heap.push(combo[i].second);
            sum += combo[i].second;
            if (heap.size() > k) {
                sum -= heap.top();
                heap.pop();
            }
            if (i >= k - 1) {
                const int64_t min_n2 = combo[i].first;
                res = std::max(res, min_n2 * sum);
            }
        }

        return res;
    }
};

File: algorithms/2_medium/M/02542_maximum_subsequence_score/heap.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

impl Solution {
    pub fn max_score(nums1: Vec<i32>, nums2: Vec<i32>, k: i32) -> i64 {
        assert!(nums1.len() == nums2.len() && k >= 1 && (k as usize) <= nums1.len());

        let len = nums1.len();
        let k = k as usize;
        let mut combo = Vec::with_capacity(len);
        unsafe { combo.set_len(len); }
        for i in 0..len {
            combo[i] = (nums2[i], nums1[i]);
        }
        combo.sort_unstable();
        let mut heap: BinaryHeap<Reverse<i32>> = BinaryHeap::with_capacity(k + 1);
        let mut sum = 0_i64;
        let mut res = i64::MIN;

        for i in (0..len).rev() {
            heap.push(Reverse(combo[i].1));
            sum += combo[i].1 as i64;
            if heap.len() > k {
                let Reverse(top) = heap.pop().unwrap();
                sum -= top as i64;
            }
            if heap.len() == k {
                let min_n2 = combo[i].0 as i64;
                res = res.max(min_n2 * sum);
            }
        }

        res
    }
}
